### DISCUSSION TITLE
Why should we use `writeFinal` when recording a final state to Vstorage

### DISCUSSION THREAD

**Discussion by Jorge-Lopes** [Agoric/agoric-sdk#8812] [Q&A] [answered]
When we retrieve a recorder object from the `makeRecorderKit` API, we get both `write` and `writeFinal` methods, with the difference being that the latest will prevent the publisher from receiving more calls. 
[recorder.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/src/contractSupport/recorder.js#L117)

```jsx
// below here differs from writeFinal()
this.state.closed = true;
return publisher.finish(value);
```

I understand that the best practice is to use writeFinal when we are executing a final call to publish the completion value. Which will prevent that node value from being updated again.

What I would like to understand in more depth is if this pattern was implemented to protect the developer in case of fault design in the contract or if it has any overload to Agoric chain to keep the Publish object listening to updates?

Thank you all in advance

**Comment by dckc** [comment]
@gibson042 ? thoughts?

**Comment by gibson042** [comment, marked as answer]
> What I would like to understand in more depth is if this pattern was implemented to protect the developer in case of fault design in the contract or if it has any overload to Agoric chain to keep the Publish object listening to updates?

The former. Our async sequence publisher pattern consists of a { publish, finish, fail } trio as described at [@agoric/notifier](https://github.com/Agoric/agoric-sdk/tree/master/packages/notifier#distributed-asynchronous-iteration) and [Notifiers and Subscriptions](https://docs.agoric.com/guides/js-programming/notifiers.html#distributed-asynchronous-iteration), and the RecorderKit interface is built on top of it. Also, finalization applies to the `publisher` object but _not_ to the chain storage node that ultimately receives its data—a distinct publisher object associated with the same chain storage node can still continue writing to it.

Is that sufficiently clear?

**Reply by Jorge-Lopes** [reply]
Yes it is @gibson042.
Thank you very much for the explanation.
